### ISSUE TITLE
QuickPulse payload reference?

### ISSUE THREAD

**Issue by rcarmo** [microsoft/ApplicationInsights-Home#315]
I'm looking for a full JSON example payload for QuickPulse, but can't seem to find it anywhere (although the Java SDK has a readable formatter that hints at the format).

**Comment by MS-TimothyMothra** [comment]
@lmolkova has context, and will reply :)

**Comment by SergeyKanzhelev** [comment]
@rcarmo this contract is not public. I.e. no guarantee on it being supported outside of shipped SDKs. What's do you want to use it for?

Here are service contracts in C# https://github.com/Microsoft/ApplicationInsights-dotnet-server/tree/3cc166e3c9ec9d3f4ce77f48d3ab8c6bd6218808/Src/PerformanceCollector/Perf.Shared.NetFull/Implementation/QuickPulse/Service%20contract

**Comment by rcarmo** [comment]
I want to use it to build LiveMetrics support for Python applications, since the SDK does not support it at present (I intend to contribute to it later on).

My current prototype uses an existing UDP metrics collector and is reproduced below. It was working a couple of months back, but something changed and I am trying to figure out why:

```python
from psutil import cpu_percent, virtual_memory
from aiohttp import ClientSession
from uvloop import EventLoopPolicy
from asyncio import set_event_loop_policy, get_event_loop, DatagramProtocol, sleep, ensure_future
from time import time
from ujson import loads, dumps
from uuid import uuid4
from socket import gethostname
from .config import INSTRUMENTATION_KEY, STATS_ADDRESS, STATS_PORT, API_KEY, log

QUICKPULSE_ID = str(uuid4()).replace("-", "");
QP_BASE_URI = "https://rt.services.visualstudio.com/QuickPulseService.svc/"
HOSTNAME = gethostname()
TICKS_AT_EPOCH = 621355968000000000

COUNTERS = {
    "requests": 0,
    "times": [],
    "failed": 0
}

class Metrics(DatagramProtocol):
    def __init__(self):
        super().__init__()

    def connection_made(self, transport):
        self.transport = transport

    def datagram_received(self, data, addr):
        event = loads(data)
        #print(event)
        COUNTERS["requests"] += 1
        COUNTERS["times"].append(event["time"])
        if event["result_code"] >= 400:
            COUNTERS["failed"] += 1


async def main(loop):
    global COUNTERS
    print("entering loop")
    gather_metrics = False
    metrics = None
    while(True):
        async with ClientSession() as session:
            if gather_metrics:
                metrics = [{
                        "Name": "\\ApplicationInsights\\Requests/Sec",
                        "Value": COUNTERS["requests"],
                        "Weight": 1
                    }, {
                        "Name": "\\ApplicationInsights\\Request Duration",
                        "Value": sum(COUNTERS["times"])/len(COUNTERS["times"]) if len(COUNTERS["times"]) else 0,
                        "Weight": 1
                    }, {
                        "Name": "\\ApplicationInsights\\Requests Failed/Sec",
                        "Value": COUNTERS["failed"],
                        "Weight": 1
                    }, {
                        "Name": "\\ApplicationInsights\\Requests Succeeded/Sec",
                        "Value": COUNTERS["requests"] - COUNTERS["failed"],
                        "Weight": 1
                    }, {
                        "Name": "\\ApplicationInsights\\Dependency Calls/Sec",
                        "Value": 0,
                        "Weight": 1
                    }, {
                        "Name": "\\ApplicationInsights\\Dependency Call Duration",
                        "Value": 0,
                        "Weight": 1
                    }, {
                        "Name": "\\ApplicationInsights\\Dependency Calls Failed/Sec",
                        "Value": 0,
                        "Weight": 1
                    }, {
                        "Name": "\\ApplicationInsights\\Dependency Calls Succeeded/Sec",
                        "Value": 0,
                        "Weight": 1
                    }, {
                        "Name": "\\ApplicationInsights\\Exceptions/Sec",
                        "Value": 0,
                        "Weight": 1
                    }, {
                        "Name": "\\Memory\\Committed Bytes",
                        "Value": virtual_memory().used,
                        "Weight": 1
                    }, {
                        "Name": "\\Processor(_Total)\\% Processor Time",
                        "Value": cpu_percent(interval=0.5),
                        "Weight": 1
                    }
                ]

            #print(metrics)
            current_time_millis = int(time() * 1000)
            payload = {
                "Documents": [],
                "Instance": HOSTNAME,
                "InstrumentationKey": INSTRUMENTATION_KEY,
                "InvariantVersion": 2,
                "MachineName": HOSTNAME,
                "StreamId": QUICKPULSE_ID,
                "Metrics": metrics,
                "Timestamp": "/Date(%d)/" % current_time_millis,
                "Version": "2.2.0-738",
            }

            #print(payload)
            async with session.post(QP_BASE_URI + "ping?ikey=" + INSTRUMENTATION_KEY, headers = {
                    "x-ms-qps-transmission-time": str(current_time_millis * 10000 + TICKS_AT_EPOCH),
                }, data = dumps(payload)) as res:

                print(payload,res, await res.text())
                if 'x-ms-qps-subscribed' in res.headers:
                    if res.headers['x-ms-qps-subscribed'].lower() == 'true':
                        print(res.headers['x-ms-qps-subscribed'])
                        gather_metrics = True
                        COUNTERS = {
                            "requests": 0,
                            "times": [],
                            "failed": 0
                        }
                        print("sleeping 1")
                        await sleep(1)
                        continue
                gather_metrics = False
                metrics = None
                print("sleeping 15")
                await sleep(15)


if __name__ == '__main__':
    #set_event_loop_policy(EventLoopPolicy())
    loop = get_event_loop()
    loop.set_debug(True)
    print("spawning main")
    ensure_future(main(loop))

    listener = loop.create_datagram_endpoint(Metrics,local_addr=(STATS_ADDRESS,STATS_PORT))
    loop.run_until_complete(listener)
    try:
        print("entering loop")
        loop.run_forever()
    finally:
        loop.close()
```

**Comment by rcarmo** [comment]
Thanks, but a peer had already looked at that, and it had trouble with asyncio in testing (not to mention being an added dependency, etc.).

**Comment by lmolkova** [comment]
@rcarmo sorry for the delay. Could you please report asyncio testing issues to OpenCensus community (https://github.com/census-instrumentation/opencensus-python)?

**Comment by rcarmo** [comment]
No problem. But we've since established that we cannot use OpenCensus for the containerised scenario I was working on. Right now either I finish the snippet I have above (which is WIP) and have basic in-container stats, or we feed those via a Redis queue to a central collector. There are technical arguments pro and against, and I'm sorting them out (among other things).

It would be a tremendous help if someone was able to share a complete JSON transaction (I looked through the C# service contracts, but re-assembling those into a payload is a time-consuming task that I have yet to finish).
